registry/storage/metadata/keys.go

```diff
@@ -0,0 +1,124 @@
+package metadata
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/reference"
+)
+
+// todo: verify key on create
+
+// Keys provides a method for generating, serializing and unserializing keys.  The
+// string representation of a key can be converted back into a strong type.  Key
+// string representations should not clash.
+func KeyFromString(keyStr string) iterable {
+	re := regexp.MustCompile("^manifest::(" + digest.DigestRegexp.String() + ")$")
```

Comment:
Define these regexps in the global scope so that they are only compiled once.

It might be a good idea to first parse out the object type and then switch on that, so you don't have to run all the parsing regexps against the input until you find one that matches.